wallet/src/wallet.rs

```diff
@@ -0,0 +1,194 @@
+        &self.db
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn add_transaction(&mut self, tx: Transaction, state: TxState) -> Result<(), WalletError> {
+        let tx_id = tx.get_id();
+
+        if self.txs.contains_key(&tx_id) {
+            return Err(WalletError::DuplicateTransaction(tx_id));
+        }
+
+        let mut db_tx = self.db.transaction_rw(None).map_err(WalletError::DatabaseError)?;
+
+        let mut wallet_tx = WalletTx::new(tx, state);
+        wallet_tx.set_order(Some(self.txs.len() as u64));
+
+        db_tx.set_transaction(&tx_id, &wallet_tx).map_err(WalletError::DatabaseError)?;
+        db_tx.commit().map_err(WalletError::DatabaseError)?;
+
+        self.txs.insert(tx_id, wallet_tx);
+
+        // TODO add UTXO?
+
+        Ok(())
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn delete_transaction(&mut self, tx_id: Id<Transaction>) -> Result<(), WalletError> {
+        if !self.txs.contains_key(&tx_id) {
+            return Err(WalletError::NoTransactionFound(tx_id));
+        }
+
+        let mut db_tx = self.db.transaction_rw(None).map_err(WalletError::DatabaseError)?;
+        db_tx.del_transaction(&tx_id).map_err(WalletError::DatabaseError)?;
+        db_tx.commit().map_err(WalletError::DatabaseError)?;
+
+        self.txs.remove(&tx_id);
+
+        // TODO remove UTXO?
+
+        Ok(())
+    }
+
+    // TODO fix incompatibility between borrowing mut self and the database transaction
+    #[allow(dead_code)] // TODO remove
+    fn add_to_utxos(
+        &mut self,
+        tx: &Transaction,
+        db_tx: &mut StoreTxRw<B>,
+    ) -> Result<(), WalletError> {
+        for (i, output) in tx.outputs().iter().enumerate() {
+            // Check if this output belongs to this wallet or it is watched
+            if self.is_available_for_spending(output) && self.is_mine_or_watched(output) {
+                let outpoint = OutPoint::new(tx.get_id().into(), i as u32);
+                let utxo = Utxo::new(output.clone(), false, utxo::UtxoSource::Mempool);
+                self.utxo.insert(outpoint.clone(), utxo.clone());
+                db_tx.set_utxo(&outpoint, utxo).map_err(WalletError::DatabaseError)?;
+            }
+        }
+        Ok(())
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn is_available_for_spending(&self, _txo: &TxOutput) -> bool {
+        // TODO implement
+        true
+    }
+
+    #[allow(dead_code)] // TODO remove
+    fn is_mine_or_watched(&self, _txo: &TxOutput) -> bool {
+        // TODO implement
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use common::chain::{GenBlock, Transaction};
+    use common::primitives::H256;
+
+    #[test]
+    fn in_memory_wallet() {
+        let wallet = open_wallet_in_memory();
+        assert!(wallet.is_ok())
+    }
+
+    #[test]
+    fn wallet_transactions() {
+        let wallet_path =
+            tempfile::TempDir::new().unwrap().path().join("wallet_transactions.sqlite");
```

Comment:
This is wrong. This is the mistake I warned from.

The object of TempDir must be held until the end of the test, because the Drop trait, or the destructor, of TempDir deletes the file. Which means, you're creating the directory and destroying it immediately. If you check your /tmp/ directory, you'll probably find all these remnants sitting there undeleted.